usersync/cookie.go

```diff
@@ -20,72 +19,148 @@ const uidTTL = 14 * 24 * time.Hour
 
 // Cookie is the cookie used in Prebid Server.
 //
-// To get an instance of this from a request, use ParseCookieFromRequest.
-// To write an instance onto a response, use SetCookieOnResponse.
+// To get an instance of this from a request, use ReadCookie.
+// To write an instance onto a response, use WriteCookie.
 type Cookie struct {
-	uids   map[string]uidWithExpiry
+	uids   map[string]UIDEntry
 	optOut bool
 }
 
-// uidWithExpiry bundles the UID with an Expiration date.
-// After the expiration, the UID is no longer valid.
-type uidWithExpiry struct {
+// UIDEntry bundles the UID with an Expiration date.
+type UIDEntry struct {
 	// UID is the ID given to a user by a particular bidder
 	UID string `json:"uid"`
 	// Expires is the time at which this UID should no longer apply.
 	Expires time.Time `json:"expires"`
 }
 
-// ParseCookieFromRequest parses the UserSyncMap from an HTTP Request.
-func ParseCookieFromRequest(r *http.Request, cookie *config.HostCookie) *Cookie {
-	if cookie.OptOutCookie.Name != "" {
-		optOutCookie, err1 := r.Cookie(cookie.OptOutCookie.Name)
-		if err1 == nil && optOutCookie.Value == cookie.OptOutCookie.Value {
-			pc := NewCookie()
-			pc.SetOptOut(true)
-			return pc
-		}
+// NewCookie returns a new empty cookie.
+func NewCookie() *Cookie {
+	return &Cookie{
+		uids: make(map[string]UIDEntry),
 	}
-	var parsed *Cookie
-	uidCookie, err2 := r.Cookie(uidCookieName)
-	if err2 == nil {
-		parsed = ParseCookie(uidCookie)
-	} else {
-		parsed = NewCookie()
+}
+
+// ReadCookie reads the cookie from the request
+func ReadCookie(r *http.Request, decoder Decoder, host *config.HostCookie) *Cookie {
+	if hostOptOutCookie := checkHostCookieOptOut(r, host); hostOptOutCookie != nil {
+		return hostOptOutCookie
+	}
+
+	// Read cookie from request
+	cookieFromRequest, err := r.Cookie(uidCookieName)
+	if err != nil {
+		return NewCookie()
 	}
-	// Fixes #582
-	if uid, _, _ := parsed.GetUID(cookie.Family); uid == "" && cookie.CookieName != "" {
-		if hostCookie, err := r.Cookie(cookie.CookieName); err == nil {
-			parsed.TrySync(cookie.Family, hostCookie.Value)
+	decodedCookie := decoder.Decode(cookieFromRequest.Value)
+
+	return decodedCookie
+}
+
+// PrepareCookieForWrite ejects UIDs as long as the cookie is too full
+func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, encoder Encoder) (string, error) {
+	uuidKeys := sortUIDs(cookie.uids)
+
+	i := 0
+	for len(cookie.uids) > 0 {
+		encodedCookie, err := encoder.Encode(cookie)
+		if err != nil {
+			return encodedCookie, nil
 		}
+
+		isCookieTooBig := len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
```

Comment:
In the previous version of this file, the cookie was trimmed according to it's `cookie.String()` size as ilustrated in line 168 below:
```
143 // SetCookieOnResponse is a shortcut for "ToHTTPCookie(); cookie.setDomain(domain); setCookie(w, cookie)"
144 func (cookie *Cookie) SetCookieOnResponse(w http.ResponseWriter, setSiteCookie bool, cfg *config.HostCookie, ttl time.Duration) {
145     httpCookie := cookie.ToHTTPCookie(ttl)
146     var domain string = cfg.Domain
147
148     if domain != "" {
149         httpCookie.Domain = domain
150     }
151
152     var currSize int = len([]byte(httpCookie.String()))
153     for cfg.MaxCookieSizeBytes > 0 && currSize > cfg.MaxCookieSizeBytes && len(cookie.uids) > 0 {
154         var oldestElem string = ""
155         var oldestDate int64 = math.MaxInt64
156         for key, value := range cookie.uids {
157             timeUntilExpiration := time.Until(value.Expires)
158             if timeUntilExpiration < time.Duration(oldestDate) {
159                 oldestElem = key
160                 oldestDate = int64(timeUntilExpiration)
161             }
162         }
163         delete(cookie.uids, oldestElem)
164         httpCookie = cookie.ToHTTPCookie(ttl)
165         if domain != "" {
166             httpCookie.Domain = domain
167         }
168         currSize = len([]byte(httpCookie.String())) // <- OLD VERSION USES httpCookie.String()
169     }
170
171     if setSiteCookie {
172         httpCookie.Secure = true
173         httpCookie.SameSite = http.SameSiteNoneMode
174     }
175     w.Header().Add("Set-Cookie", httpCookie.String())
176 }
usersync/cookie.go
```
To be consistent, the refactored version could also make use of it:
```
61   func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, encoder Encoder) (string, error) {
62       uuidKeys := sortUIDs(cookie.uids)
63  
64       i := 0
65       for len(cookie.uids) > 0 {
66           encodedCookie, err := encoder.Encode(cookie)
67           if err != nil {
68               return encodedCookie, nil
69           }
   +         
   +         oldFashionedCookieSize := len([]byte(&http.Cookie{
   +             Name:    uidCookieName,
   +             Value:   b64,
   +             Expires: time.Now().Add(ttl),
   +             Path:    "/",
   +         }).String())
70  
71 -         isCookieTooBig := len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
   +         isCookieTooBig := len(oldFashionedCookieSize) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
72           if !isCookieTooBig {
73               return encodedCookie, nil
74           }
75  
76           uidToDelete := uuidKeys[i]
77           delete(cookie.uids, uidToDelete)
78  
79           i++
80       }
81       return "", nil
82   }
usersync/cookie.go
```
Back when I coded PR #2819 I was split on this. I couldn't decide on whether or not to keep using `httpCookie.String()` because it seemed wasteful and thought about simply using the base 64 encoded value like you are doing in this PR. But, after some tests, I noticed I wasn't getting the same results and decided to keep using `httpCookie.String()` in an effort to not risk changing the outcome with respect to the current version in production.

Why did you decide to not use the `httpCookie.String()` method anymore? Was this discussed offline?

Reply:
This is a good question! I decided to just use the base 64 encoded value since I assumed that that value represented the entirety of the cookie. However, you make a good point that this value eventually gets added to an `HttpCookie` and thus the size of `base64` vs. `httpCookie.String()` are different. I think it makes sense that we utilize the `oldFashionedCookieSize`, but it's something that I'll bring up in team time to confirm.